Threads: Read receipts & notifications (#1255)
matrix-org/matrix-spec: content/client-server-api/modules/receipts.md

[Receiving notifications](#receiving-notifications) section for more
information on how read receipts affect notification counts.

> {{< added-in v="1.4" >}} Read receipts exist in three major forms:

Reviewer: MSC3771 doesn't specify that unthreaded/threaded applies only to *read* receipts, but there isn't currently any other type of receipts...

Author: yea, we're very slowly moving this module to talk about only read receipts at the moment (at least until another receipt shows up)

Reviewer: Makes sense, just wanted to ensure this was on purpose.

receipt acknowledgements based on `user_id`, `receipt_type`, and the
`thread_id` (if present).
For example:

    Client receives m.receipt:
      user = @alice:example.com
      receipt_type = m.read
      event_id = $aaa:example.com
      thread_id = undefined

    Client receives another m.receipt:
      user = @alice:example.com
      receipt_type = m.read
      event_id = $bbb:example.com
      thread_id = main

    The client does not replace any acknowledgements, yet.

    Client receives yet another m.receipt:
      user = @alice:example.com
      receipt_type = m.read
      event_id = $ccc:example.com
      thread_id = undefined

    The client replaces the older acknowledgement for $aaa:example.com
    with this new one for $ccc:example.com, but does not replace the
    acknowledgement for $bbb:example.com because it belongs to a thread.

    Client receives yet another m.receipt:
      user = @alice:example.com
      receipt_type = m.read
      event_id = $ddd:example.com
      thread_id = main

    Now the client replaces the older $bbb:example.com acknowledgement with
    this new $ddd:example.com acknowledgement. The client does NOT replace the

>     older acknowledgement for $ccc:example.com as it is unthreaded.

Reviewer 2: Do we not need to update some language about how clients should *render* unthreaded vs threaded receipts? E.g. how should unthreaded vs threaded clients treat  receiving unthreaded vs threaded receipts?

Author: oh, that would probably be good, yes. Will add a sentence somewhere to say that.

uncommon, it is considered valid to have an `m.read` (public) receipt lag
several messages behind the `m.read.private` receipt, for example.

##### Threaded read receipts

{{% added-in v="1.4" %}}

If a client does not use [threading](#threading), then they will simply only
send "unthreaded" read receipts which affect the whole room regardless of threads.

Threading introduces a concept of multiple conversations being held in the same
room and thus deserve their own read receipts and notification counts. An event
is considered to be "in a thread" if it has either a `rel_type` of `m.thread` or
has child events with a `rel_type` of `m.thread` (in which case it'd be the
thread root). Events not in a thread but still in the room are considered to be

> part of the "main timeline", or a special thread with an ID of `main`.

Reviewer: Should this have similar language [to above](https://github.com/matrix-org/matrix-spec/pull/1255/files#diff-bfda8c579b08f5dd97e312b0a592762a36737a8477e23dc9273936eb9b2d6aebR126) discussing going 3 hops to find the thread?

Author: possibly, though it's not immediately obvious that it's what the MSC intended - will dig a bit deeper on it.

Reviewer: Both MSC3771 and MSC3773 have similar language like:

> It is recommended that at least 3 relations are traversed when attempting to find a thread, implementations should be careful to not infinitely recurse.

This is for both ensuring the receipt refers to a valid event ID & when partitioning a push action by thread.

> A threaded read receipt is simply one which has a `thread_id` on it, targeting

Reviewer 3: I would probably move that paragraph 1 level up so that it follows the reference of "unthreaded". As it is mentioned, but not defined.
This paragraph addresses that

either a thread root's event ID or `main` for the main timeline.

The following is an example DAG for a room, with dotted lines showing event
relationships and solid lines showing topological ordering.

![threaded-dag](/diagrams/threaded-dag.png)

{{% boxes/note %}}
`m.reaction` relationships are not currently specified, but are shown here for
their conceptual place in a threaded DAG. They are currently proposed as

> [MSC2677](https://github.com/matrix-org/matrix-spec-proposals/pull/2677).

Reviewer: You could replace them with `m.reference` relations if #1206 were to be merged. 🤷

Author: `m.reaction` is still clearer imo

{{% /boxes/note %}}

This DAG can be represented as 3 threaded timelines, with `A` and `B` being thread
roots:

![threaded-dag-threads](/diagrams/threaded-dag-threads.png)

With this, we can demonstrate that:
* A threaded read receipt on `I` would mark `A`, `B`, and `I` as read.
* A threaded read receipt on `E` would mark `C` and `E` as read.
* An unthreaded read receipt on `D` would mark `A`, `B`, `C`, and `D` as read.

Note that marking `A` as read with a threaded read receipt would not mean
that `C`, `E`, `G`, or `H` get marked as read: Thread A's timeline would need
its own threaded read receipt at `H` to accomplish that.

The read receipts for the above 3 examples would be:

> ```json

Reviewer 3: Same comment as the other spec proposal. Should this use `json5` or `jsonc` ?

{
  "$I": {
    "m.read": {
      "@user:example.org": {
        "ts": 1661384801651,
        "thread_id": "main" // because `I` is not in a thread, but is a threaded receipt
      }
    }
  },
  "$E": {
    "m.read": {
      "@user:example.org": {
        "ts": 1661384801651,
        "thread_id": "$A" // because `E` is in Thread `A`
      }
    }
  },
  "$D": {
    "m.read": {
      "@user:example.org": {
        "ts": 1661384801651
        // no `thread_id` because the receipt is *unthreaded*
      }
    }
  }
}
```

> Conditions on sending read receipts, such as using private read receipts, only
> sending once an event is display, not being able to move backwards in time, etc
> still apply to threaded and unthreaded read receipts. For example, a client might
> send a private read receipt for a threaded event when the user expands that thread.

Reviewer 3: I found that relatively difficult to parse. The length of examples of condition did hide the important bit of the sentence. What do you think of the following? 

```suggestion
Conditions on sending read receipts apply similarly to threaded and unthreaded read receipts. For example, a client might
send a private read receipt for a threaded event when the user expands that thread.
```

Author: have applied manually - thanks!

{
    <room_id>: {
        <receipt_type>: {

>             <user_id>: { <content (ts & thread_id, currently)> }

Reviewer: I suspect there should be corresponding updates to the [server-server API for receipts](https://pr1255--matrix-spec-previews.netlify.app/server-server-api/#mreceipt), at least to "Read Receipt Metadata".
